app/src/main/java/org/mozilla/fenix/components/metrics/LeanplumMetricsService.kt

```diff
@@ -58,46 +61,52 @@ class LeanplumMetricsService(private val application: Application) : MetricsServ
-            true -> applicationSetLocale.isO3Language
-            false -> Locale.getDefault().isO3Language
-        }
-        if (!isLeanplumEnabled(currentLocale)) {
-            Log.i(LOGTAG, "Leanplum is not available for this locale: $currentLocale")
-            return
-        }
+            val applicationSetLocale = LocaleManager.getCurrentLocale(application)
+            val currentLocale = when (applicationSetLocale != null) {
+                true -> applicationSetLocale.isO3Language
+                false -> Locale.getDefault().isO3Language
+            }
+            if (!isLeanplumEnabled(currentLocale)) {
+                Log.i(LOGTAG, "Leanplum is not available for this locale: $currentLocale")
+                return@launch
+            }
 
-        when (token.type) {
-            Token.Type.Production -> Leanplum.setAppIdForProductionMode(token.id, token.token)
-            Token.Type.Development -> Leanplum.setAppIdForDevelopmentMode(token.id, token.token)
-            Token.Type.Invalid -> {
-                Log.i(LOGTAG, "Invalid or missing Leanplum token")
-                return
+            when (token.type) {
+                Token.Type.Production -> Leanplum.setAppIdForProductionMode(token.id, token.token)
+                Token.Type.Development -> Leanplum.setAppIdForDevelopmentMode(token.id, token.token)
+                Token.Type.Invalid -> {
+                    Log.i(LOGTAG, "Invalid or missing Leanplum token")
+                    return@launch
+                }
             }
-        }
 
-        Leanplum.setIsTestModeEnabled(false)
-        Leanplum.setApplicationContext(application)
-        Leanplum.setDeviceId(randomUUID().toString())
-        Parser.parseVariables(application)
-
-        LeanplumActivityHelper.enableLifecycleCallbacks(application)
-
-        val installedApps = MozillaProductDetector.getInstalledMozillaProducts(application)
-
-        Leanplum.start(application, hashMapOf(
-            "default_browser" to MozillaProductDetector.getMozillaBrowserDefault(application).orEmpty(),
-            "fennec_installed" to installedApps.contains(MozillaProductDetector.MozillaProducts.FIREFOX.productName),
-            "focus_installed" to installedApps.contains(MozillaProductDetector.MozillaProducts.FOCUS.productName),
-            "klar_installed" to installedApps.contains(MozillaProductDetector.MozillaProducts.KLAR.productName),
-            "fxa_signed_in" to application.settings().fxaSignedIn,
-            "fxa_has_synced_items" to application.settings().fxaHasSyncedItems,
-            "search_widget_installed" to application.settings().searchWidgetInstalled,
-            "fenix" to true
-        ))
+            Leanplum.setIsTestModeEnabled(false)
+            Leanplum.setApplicationContext(application)
+            Leanplum.setDeviceId(randomUUID().toString())
+            Parser.parseVariables(application)
+
+            LeanplumActivityHelper.enableLifecycleCallbacks(application)
```

Comment:
This one is more concerning as it sets a `ActivityLifecycleCallback`, which receives callbacks on the main thread. It in turn invokes `onPause`/`onResume`/`onStop` on `Leanplum`, which reads/writes state that isn't thread-safe either. For some of these states it doesn't matter as they are only written/read on the main thread e.g. `isActivityPaused` and `currentActivity`. However, `pause` also checks `hasCalledStart` which is now set on an IO thread when starting `Leanplum`. Since, there's no visibility guarantee for that field, it could be that we fail to pause properly. As far as I can tell, this could cause us to continue to send heartbeats while paused. I think we need to discuss this more and do some manual testing.

(Sorry if you get this twice, Github somehow dropped my earlier comment)

Reply:
> Since, there's no visibility guarantee for that field

For a short term quick fix, could we fork the SDK and add `@Volatile`?

Reply:
I think if `calledStart` is our only concern we could also switch to the MainScope() after we successfully started and set it ourselves (on the right thread). What we have to do, I think, is go over all states set by the lifecylce callback and make sure we understand the consequences. It seems doable to review those states.